### PR TITLE
Using viper for config management - WIP

### PR DESCRIPTION
**Pull request by tsheasha**
This means on config change we only need to reload config for the
collector instead of restarting the whole fullerite service

### CODE REVIEW

File: src/fullerite/collectors.go [hunk 1]
```diff
@@ -9,25 +9,30 @@ import (
...
 
 	for _, name := range c.Collectors {
-		configFile := strings.Join([]string{c.CollectorsConfigPath, name}, "/") + ".conf"
+		conf := v.New()
+		conf.SetConfigType("json")
 		// Since collector naems can be defined with a space in order to instantiate multiple
 		// instances of the same collector, we want their files
 		// will not have that space and needs to have it replaced with an underscore
 		// instead
-		configFile = strings.Replace(configFile, " ", "_", -1)
```
Comment (tiras-j) [inline src/fullerite/collectors.go]: Do we want to leave spaces in the config names or replace them with _?

Comment (tsheasha) [inline src/fullerite/collectors.go]: replace with _ is more sane imho

File: src/fullerite/collectors.go [hunk 2]
```diff
@@ -47,6 +52,10 @@ func startCollector(name string, globalConfig config.Config, instanceConfig map[
 
 	// apply the instance configs
 	collectorInst.Configure(instanceConfig)
+	instanceConfig.WatchConfig()
+	instanceConfig.OnConfigChange(func(e fsnotify.Event) {
+		collectorInst.Configure(instanceConfig)
+	})
 
 	go runCollector(collectorInst)
```
Comment (gnufied) [inline src/fullerite/collectors.go]: I hope all collector configuration setting is thread safe and re-entrat because basically, what we will be doing is while they are running we are going to change their configuration underneath.

**Comment by gnufied** [comment]
Correct me if I am wrong, but this does not cover fullerite core configuration and neither diamond collectors right? This only covers configuration of collectors written in go?

File: src/fullerite/collectors.go (hunk 2 above)
Comment (tsheasha) [inline src/fullerite/collectors.go]: yes that is the point * to change their config while running *. Collectors should not **change** any fields set by the config. They only need to read from it, I have checked the current ones and none of them does it differently

**Comment by tsheasha** [comment]
@gnufied well since viper is a go library it can not support python collector configuration management.

As for the core config file, on changing that file it is more sane to restart all collectors and handlers and instead of doing that within the application, I would rather do that by restarting the service.

File: src/fullerite/collectors.go (hunk 1 above)
Comment (tiras-j) [inline src/fullerite/collectors.go]: Don't we need to add the Replace back in then? It's gone with this change isn't it?

Comment (tsheasha) [inline src/fullerite/collectors.go]: fixed

**Comment by tiras-j** [comment]
lgtm!
+1

**Comment by gnufied** [comment]
With due respect, this pull request comes with some drawbacks and introduces more moving parts for fixing a problem, we didn't had in first place.  For example, we haven't had lost data  or any problem that I know of - because fullerite had to be restarted so as a collector can re-read its config file.

On other hand, the thread safety problems are real, even if I see how things will become eventually consistent. But my problem is, collector for awhile might be running half of old config and half of new and also contract between collector and what it can do with its config data will be more rigid (thou shalt not modify your config. data in collector, because they will be lost).


In a nutshell, I think fullerite is relatively lightweight and startsup pretty quickly. I don't think we should worry about automatically re-reading config files, given the cost it imposes.

**Comment by tsheasha** [comment]
1) This PR is providing an enhancement, so there is no problem read bug ) we have currently that I am fixing

2) Thread safety will at worst case affect only one collection, which will be actually the same as if you restart the service you still lose one collection. As far as the contract goes, if someone is changing values they receive from config in the program code then I am sorry to say **they are doing it wrong**. Config in fullerite is static and not to be tampered with outside of your config management software. So in case someone creates s collector that tampers with the config, that person needs to be advised against it whether or not we use viper.

3) automatically reloading config is very useful to avoide ibterrupting collections that ca  be holding http connections or mysql connections for exampke etc...

So in a nutshell i am pro merging this and i disagree with the provided arguments against it

**Comment by zachm** [comment]
I tend to agree with @gnufied that this PR falls under YAGNI. Reasons why:
- We rarely, rarely, rarely change any kind of configs.
- Especially with the adhoc collector being available, frequency of changes will decrease even further.
- The thread safety issues make me nervous.
- This muddles the messaging. You need to restart Fullerite to pick up config changes, with the core and with diamond. Folks writing Go collectors will be much less common than just turning on extant Diamond collectors. Then even Go collector folks will trigger a restart anyway, "just to be safe" and we're back at square one.

On the other hand I see @tsheasha putting forth great effort on behalf of progress. I think I could be persuaded to +1 if we (a) added some sort of protection for the "doing it wrong" scenario, (b) implemented hot configs for the core, (c) implemented some sort of transparent hot configs for diamond collectors. That said, my requests are undoubtedly more trouble than they're worth.

**Comment by tsheasha** [comment]
@zachm to answer your points:-
1) Yes *we* rarely change them, however give the open source nature of this project other people using it change them quite frequently, I know of at least two people using this that have passwords being passed to the config that are changed on a 6 hour basis.
2) I do not see how the adhoc collector will reduce config change to a persistent collector, can you please elaborate? A config change is usually done using a config management software to run *everywhere* so using an adhoc collector to *test* the new config might be well and good, otherwise, it don't really see how it is one or the other?
3) The thread safety issues are actually negligible, since only one source is changing the config files so we will never need to hold an "exclusive lock" on a resource, worst case we get stale data for one collection run, which will happen only for one collector.
4) Diamond already implements a reactive config change mechanism using SIGHUP signals we just need to *activate* it this follows point(c) in your last paragraph which is something I can indeed add so +1 for that
5) adding protection for "doing it wrong" scenario is actually impossible to do given that each collector has an arbitrary number of instance variables, that are assigned by the config. This issue exists anyways with or without viper, since even if we use the old config class if someone *wants* to change the config they can, they will just be *doing it wrong* and hence the code review process.

(a) addressed in point # 5
(b) and (c) sounds like a very easy task and very doable will include them tomorrow

**Comment by chrisenuf** [comment]
Regarding the thread safety concerns, can you set the config change watcher to update the configuration on a delay?  It wouldn't really solve the issue, but it would make it very unlikely that the collector is reading in values that are in the middle of being changed.

If you can address ztm's (b) and (c), I don't really have any problem with this.

**Comment by gnufied** [comment]
Go 1.6 by the looks of it, makes a change which will make the program crash if two goroutines are reading/writing on same map. https://golang.org/doc/go1.6 . 

I haven't looked underneath viper but I assume it uses maps to store key/value pairs somewhere.

**Comment by tiras-j** [comment]
@gnufied fwiw I see both sides of this argument as reasonable, but I do want to point out that the viper config object doesn't notify on a watch until it has entirely read in the underlying file, so there isn't a risk of concurrent write/read of the underlying map. Relevant code: https://github.com/spf13/viper/blob/master/viper.go#L232

**Comment by tsheasha** [comment]
@chrisenuf I will address (b) and (c) then

@gnufied the map is never read from or written to concurrently so I am not sure how go 1.6 will affect us. The map is updated on file change as @tiras-j mentioned and the map is filled. Then we call configure, which takes that map after it has been filled fully and use it to change the values of the instance variables of the specific collector struct. So the race condition is **not** reading and writing to the config underlying map it is rather when we try to access attributes of the collector like for example if the config changes the `interval` of collection, this change will take place after we call configure and will be effective from the next collection round. 

Basically the config map itself is only written to and read from in one place and it is done sequentially **not** concurrently, so the race condition is not there

**Comment by tiras-j** [comment]
Also, if we really want to make certain that a collect() doesn't happen while a reconfigure occurs, we could simply pause the collection goroutine while the configure is happening. Something like:

```go
const (
    Running = 0
    Paused = 1
)
...
        // apply the instance configs
        s := make(chan int)
	collectorInst.Configure(instanceConfig)
	instanceConfig.WatchConfig()
	instanceConfig.OnConfigChange(func(e fsnotify.Event) {
                s <- Paused
		collectorInst.Configure(instanceConfig)
                s <- Running
	})

	go runCollector(collectorInst, s)
	return collectorInst
...
func runCollector(collector collector.Collector, s <-chan int) {
        state := Running
	log.Info("Running ", collector)

	ticker := time.NewTicker(time.Duration(collector.Interval()) * time.Second)
	collect := ticker.C

	staggerValue := 1
	collectionDeadline := time.Duration(collector.Interval() + staggerValue)

	for {
		select {
                case state = <-s: break
		case <-collect:
                        if state == Paused { break }
			if collector.CollectorType() == "listener" {
				collector.Collect()
			} else {
				countdownTimer := time.AfterFunc(collectionDeadline*time.Second, func() {
					reportCollector(collector)
				})
				collector.Collect()
				countdownTimer.Stop()
			}
		}
	}
	ticker.Stop()
}
```

Not sure if it'll work but it's fairly straightforward.

**Comment by tsheasha** [comment]
@tiras-j that sounds like a good idea, something similar might work, I will get on it and push a patch tomorrow

**Comment by gnufied** [comment]
@tiras-j @tsheasha I stand corrected about viper itself being thread safe. thank you for checking viper code for that. 

My earlier concern about collector config. itself being problematic stands. The problem is two fold:

1. Setting collector config. values should be thread safe. 
2. `Configure` method in collector should be idempotent.

To give an example, lets take `Mesos` collector where calling `Configure` on Collector results in Collector picking a new leader - https://github.com/Yelp/fullerite/blob/master/src/fullerite/collector/mesos.go#L104 . When we call `Configure` from a goroutine other than Collector goroutine and it proceeds to modify collector's leader, what collector goroutine may see is - a **corrupt** mesos leader instance, because it is possible that leader has been initialized (via barebones `new`) but hasn't been configured - https://github.com/Yelp/fullerite/blob/master/src/fullerite/collector/mesos.go#L110 

Now, it depends on collector how it deals with partially initialized structs, but it is not a mere data race problem that will fix itself on next collection. The collector might crash fullerite too, if it can't deal with those error conditions.

**Comment by tsheasha** [comment]
As I said this is a WIP, and the data race will be fixed.

**Comment by tsheasha** [comment]
As for a collector crashing fullerite if it fails, I find that highly improbable, since non of the errors here will make the collector panic afaict, anyways, this will not exist any longer after my coming patch so I will avoid discussing this further

**Comment by tsheasha** [comment]
After spending some more time on this, it is really too much effort to be spent for such a change, hence I am dropping this PR.
